safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,208 @@
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cors
+
+import (
+	"net/textproto"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-safeweb/safehttp"
+)
+
+const customHeader string = "X-Cors"
+
+var disallowedMethods = map[string]bool{
+	safehttp.MethodHead: true,
+	safehttp.MethodGet:  true,
+	safehttp.MethodPost: true,
+}
+
+var disallowedContentTypes = map[string]bool{
+	"application/x-www-form-urlencoded": true,
+	"multipart/form-data":               true,
+	"text/plain":                        true,
+}
+
+// Interceptor handles CORS requests based on its settings.
+//
+// For more info about CORS, see: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
+//
+// Constraints
+//
+// - The request methods GET, HEAD and POST are banned and will result in a 405
+//   Method Not Allowed response.
+// - The content types "application/x-www-form-urlencoded", "multipart/form-data"
+//   and "text/plain" are also banned and will result in a  415 Unsupported Media
+//   Type response.
+// - Each CORS request must contain the header "X-Cors: 1" or "Sec-Fetch-Mode: cors".
+// All of this is to prevent XSRF.
+type Interceptor struct {
+	// AllowedOrigins determines which origins should be allowed in the
+	// Access-Control-Allow-Origin header.
+	AllowedOrigins map[string]bool
+	allowedHeaders map[string]bool
+	// ExposedHeaders determines which headers should be set in the
+	// Access-Control-Expose-Headers header. If ExposedHeaders is nil, then the
+	// header is not set.
+	ExposedHeaders []string
+	// AllowCredentials determines if Access-Control-Allow-Credentials should be
+	// set to true when requests containing Cookie headers arrive.
+	AllowCredentials bool
+	// MaxAge sets the Access-Control-Max-Age header.
+	// MaxAge=0 results in MaxAge: 5.
+	MaxAge int
+}
+
+// Default creates a CORS Interceptor with default settings.
+// Those defaults are:
+//   - No Exposed Headers
+//   - No Allowed Headers
+//   - AllowCredentials: false
+//   - MaxAge: 5 seconds
+func Default(allowedOrigins ...string) *Interceptor {
+	ao := map[string]bool{}
+	for _, o := range allowedOrigins {
+		ao[o] = true
+	}
+	return &Interceptor{
+		AllowedOrigins: ao,
+		allowedHeaders: map[string]bool{},
+	}
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
```

Comment:
Why `"*"`?

And why `"X-Cors"`? It will always be an allowed header by default. Letting users specify it in `allowedHeaders` doesn't make a difference, right?